### PR TITLE
WIP File Driven Commands

### PR DESCRIPTION
**Pull request by arafatkatze**
Original Issue -> https://github.com/sourcegraph/cody/issues/588 
Original PR (Thanks to @umpox ) -> https://github.com/sourcegraph/cody/pull/584 

### Explanation of the problem 
`Explain`|`Refactor`|`Document`|`Test` are special cases for commands. They should not open the sidebar. They should instead open up with a default prompt in command pallet or inline chat depending on the context.  Based on the use case, the user can tweak the default prompt for their own purposes. We shouldn't change these commands in the sidebar, but in the command palette and in inline chat for files we should provide a lot more control.

## Explain
Explain code gives you the ability to explain the selected code in an inline chat. The inline chat comes pre-filled with a [question](https://github.com/sourcegraph/cody/pull/682/files#diff-32a4361723bc7dfefc30c08569d8ce211c86809fc957dcfec88387ae1f2c5d72R240) "Explain the selected code" 
1. Shortcut Menu
image
2. Inline chat
image

3. Result of the inline chat
image

## Generate Documentation
Here it comes pre-filled with the prompt to generate documentation and the fixup can be made depending on the user's approval of the fixup task

1. Shortcut Menu
image

2. Fixup Result
image

3. Fixup Diff
image

## Generate test
Explain code gives you the ability to explain the selected code in an inline chat. The inline chat comes pre-filled with a [question](https://github.com/sourcegraph/cody/pull/682/files#diff-32a4361723bc7dfefc30c08569d8ce211c86809fc957dcfec88387ae1f2c5d72R248) "Generate a unit test for this code" 

The user can change this to something else to generate different kinds of tests

1. Inline chat
image

2. Result of a mildly modified Inline chat
image

### Refactor code
Here it comes pre-filled with the prompt to generate refactoring and the fixup can be made depending on the user's approval of the fixup task

1. Command pallet menu 
image

2. Fixup Result
image

@umpox @abeatrix  @toolmantim  Right now this made the most sense to me as a first hand user. I want to understand if this makes sense from a product standpoint and if it does then I can clean up stuff, add some tests and make a proper PR. Alternatively, I am happy to modify these functions based on the feedback received. 

## Test plan
- Tested Locally
- Need the approval on the overall strategy of the PR before I design testing and clean up potentially redundant code.

### CODE REVIEW

File: vscode/src/custom-prompts/utils/menu.ts [hunk 1]
```diff
@@ -10,8 +10,11 @@ export const NewCustomCommandConfigMenuOptions = {
 }
 
 const inlineSeparator: QuickPickItem = { kind: -1, label: 'inline' }
-const chatOption: QuickPickItem = { label: 'Ask a Question' }
+const chatOption: QuickPickItem = { label: 'Explain code' }
```
Comment (abeatrix) [inline vscode/src/custom-prompts/utils/menu.ts:13]:
```suggestion
const chatOption: QuickPickItem = { label: 'Ask a Question' }
```
the ''Ask a Question'' option is for creating a new inline chat from the menu

File: vscode/src/custom-prompts/utils/menu.ts [hunk 2]
```diff
@@ -10,8 +10,11 @@ export const NewCustomCommandConfigMenuOptions = {
...
 const fixOption: QuickPickItem = { label: 'Refactor This Code' }
+const docOption: QuickPickItem = { label: 'Generate Documentation' }
```
Comment (abeatrix) [inline vscode/src/custom-prompts/utils/menu.ts:15]: i think these new refactor options should be added to the Refactor menu, as this is the menu for custom commands🤔

File: vscode/src/custom-prompts/CommandsController.ts [hunk 3]
```diff
@@ -230,9 +237,15 @@ export class CommandsController implements VsCodeCommandsController, vscode.Disp
                 case selectedCommandID === menu_options.config.label:
                     return await vscode.commands.executeCommand('cody.settings.commands')
                 case selectedCommandID === menu_options.chat.label:
-                    return await vscode.commands.executeCommand('cody.inline.new')
+                    return await vscode.commands.executeCommand('cody.inline.new', 'Explain the selected code')
```
Comment (abeatrix) [inline vscode/src/custom-prompts/CommandsController.ts:240]:
```suggestion
                    return await vscode.commands.executeCommand('cody.inline.new')
```

this is used for starting a new empty inline chat. if user wants to run the explain command, they can enter `/explain` so i don't think we will need to pre-fill this bit for them. what do you think?

**Review by abeatrix — CHANGES_REQUESTED** [review]
Thank you so much for initiating the changes! There might be a misunderstanding, as we still want to keep the default commands to work as /commands in chat, so i don't think any of the commands should be removed from the default command list unless there was another plan that I am not aware of. 😅 

@toolmantim are we going to replace the default /doc and /test commands with refactor commands? Or are we keeping both? 

Edit: I have also replied to the original PR as there are some moving pieces that might be worth waiting until they are finalized, sorry for the trouble caused 😞

**Comment by arafatkatze** [comment]
@abeatrix Thank you for the detailed update, and I completely respect the need to align with design requirements before proceeding with the PR, so it's okay to wait. 

I'm genuinely fascinated by Cody and am eager to delve deeper into how it works by making meaningful contributions. In fact, I've earmarked some time this week to work on it.

Could you kindly guide me to an issue that would be suitable for someone looking to understand Cody better and contribute simultaneously? I noticed this list of issues (https://github.com/sourcegraph/cody/issues/582), but if there's anything specific you think would align well with my intention to understand Cody's working while contributing on it, then I'd be happy to take it on. 

Thank you!

**Comment by abeatrix** [comment]
@arafatkatze im so sorry I missed your message! If you're still interested, we have a backlog of issues in the Cody project board that you can take a look and see if you find anything sparks your interests. (Might worth asking in the issue to confirm if it is still relevant before starting any works.)
Also, feel free to ping me anytime you have questions when working on an issue, always happy to help 😃

**Comment by arafatkatze** [comment]
This issue has been fixed by the followup prs to this

**Pull request closed without merging**